alternate approach to transaction validation

stringhandler opened the pull request.

philipr-za commented on base_layer/wallet/migrations/2021-07-05-13201407_metadata_signature/up.sql:
```diff
@@ -4,15 +4,15 @@
-    id INTEGER NOT NULL PRIMARY KEY,
-    commitment BLOB NOT NULL,
+    id INTEGER NOT NULL PRIMARY KEY,  --auto inc,
+    commitment BLOB NULL,
```
Why remove this constraint?

stringhandler replied:
The schema.rs file has this as nullable, so this makes the DB align with the current rust and schema.rs. Fixing the rust is the correct place to do this, but was out of the scope of this already large PR

philipr-za commented on base_layer/wallet/migrations/2021-07-05-13201407_metadata_signature/up.sql:
```diff
@@ -4,15 +4,15 @@
+    value BIGINT NOT NULL,
     flags INTEGER NOT NULL,
-    maturity INTEGER NOT NULL,
+    maturity bigint NOT NULL,
```
```suggestion
    maturity BIGINT NOT NULL,
```

philipr-za commented on base_layer/wallet/migrations/2021-07-05-13201407_metadata_signature/up.sql:
```diff
@@ -4,15 +4,15 @@
     status INTEGER NOT NULL,
-    tx_id INTEGER NULL,
-    hash BLOB NOT NULL,
+    tx_id bigint NULL,
```
```suggestion
    tx_id BIGINT NULL,
```

philipr-za commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -162,10 +162,25 @@ where TBackend: OutputManagerBackend + 'static
-
+        // Should probably be block time or at least configurable
+        // Start a bit later so that the wallet can start up
+        let mut txo_validation_interval = interval_at(
```
Should def be configurable. 

My next question is does this process have to happen on the regular? It is going to thrash the battery and surely doesn't have to happen at such a high rate for output validation?

If we are going with this automagic approach we should at least allow the client to control the frequency of these polling events like we do with the power_mode in the transaction service to be able to lower the polling interval and save battery.

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/database.rs:
```diff
@@ -50,12 +50,36 @@ const LOG_TARGET: &str = "wallet::output_manager_service::database";
     /// `outputs_to_be_received` from a `PendingTransactionOutputs` record into the `unspent_outputs` and
     /// `spent_outputs` collections.
-    fn confirm_transaction(&self, tx_id: TxId) -> Result<(), OutputManagerStorageError>;
+    fn confirm_transaction_encumberance(&self, tx_id: TxId) -> Result<(), OutputManagerStorageError>;
```
Nit: I don't really agree with the change to this name. Confirming the transaction actually removes encumberance.

stringhandler replied:
The idea here was to distinguish it from the 'Mined confirmed'. I think it was originally used when a transaction was confirmed, but I need to have the mined_height and hash to confirm it now. Let me have  a look and see if I can add it

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -305,6 +339,7 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
                     new_output.commit(&(*conn))?
                 },
                 DbKeyValuePair::PendingTransactionOutputs(tx_id, p) => {
+                    // TODO: Used only by coinbases, so should be renamed
```
This is used to keep track of spendable outputs that have been encumbered against a specific tx_id. Do you plan to do that in a different way because the logic to encumber outputs has not been changed in this PR thus far? If it is changing then I why would Coinbases be different?

stringhandler replied:
It's actually never set anywhere in the code at present, so the code would never run. EncumberedToBeSpent was updated by a different process and still is

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -1531,6 +1692,29 @@ impl Encryptable<Aes256Gcm> for KeyManagerStateSql {
+    }
+
+    fn decrypt(&mut self, _cipher: &Aes256Gcm) -> Result<(), Error> {
+        unimplemented!("Not supported")
```
So I agree we don't need to access the master key or branch seed yet while we only support a single wallet but if we ever have a secondary branch seed for a child wallet we will have to put this back.

philipr-za commented on base_layer/wallet/src/output_manager_service/tasks/txo_validation_task_v2.rs:
```diff
@@ -0,0 +1,420 @@
+                .get_last_spent_output()
+                .await
+                .for_protocol(self.operation_id)
+                .unwrap()
```
unwrap

philipr-za commented on base_layer/wallet/src/output_manager_service/tasks/txo_validation_task_v2.rs:
```diff
@@ -0,0 +1,420 @@
+                    .get_base_node_block_at_height(mined_height, client)
+                    .await
+                    .for_protocol(self.operation_id)?;
+                if block_at_height.is_none() || block_at_height.unwrap() != mined_in_block_hash {
```
```suggestion
                if block_at_height != Some(mined_in_block_hash) {
```

philipr-za commented on base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs:
```diff
@@ -497,56 +482,70 @@ where TBackend: TransactionBackend + 'static
-                "Transaction (TxId: {}) detected as mined but UNCONFIRMED with {} confirmations",
-                self.tx_id,
-                response.confirmations
+                "Broadcast transaction detected as mined, will be managed by transaction validation protoocol"
```
```suggestion
                "Broadcast transaction detected as mined, will be managed by transaction validation protocol"
```

philipr-za commented on base_layer/wallet/src/transaction_service/protocols/transaction_validation_protocol_v2.rs:
```diff
@@ -0,0 +1,417 @@
+        Self {
+            operation_id: 122, // Get a real tx id
+            db,
+            batch_size: 10,
```
Should come from config

philipr-za commented on base_layer/wallet/src/transaction_service/protocols/transaction_validation_protocol_v2.rs:
```diff
@@ -0,0 +1,417 @@
+                .get_last_mined_transaction()
+                .await
+                .for_protocol(self.operation_id)
+                .unwrap()
```
unwrap

philipr-za commented on base_layer/wallet/src/transaction_service/protocols/transaction_validation_protocol_v2.rs:
```diff
@@ -0,0 +1,417 @@
+                    .get_base_node_block_at_height(mined_height, client)
+                    .await
+                    .for_protocol(self.operation_id)?;
+                if block_at_height.is_none() || block_at_height.unwrap() != mined_in_block_hash {
```
```suggestion
                if block_at_height != Some(mined_in_block_hash) {
```

SWvheerden commented on base_layer/core/src/base_node/rpc/service.rs:
```diff
@@ -308,6 +325,112 @@ impl<B: BlockchainBackend + 'static> BaseNodeWalletService for BaseNodeWalletRpc
+
+        let deleted_bitmap = self
+            .db
+            .fetch_complete_deleted_bitmap_at(metadata.best_block().clone())
```
This is a note, this can be an expensive call to run if the agent requestion this call keeps asking for a header way back in the past

SWvheerden commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -350,24 +365,53 @@ where TBackend: OutputManagerBackend + 'static
+                //     }
+                // });
+
+                let utxo_validation = TxoValidationTaskV2::new(
```
nit: I would keep this as `TxoValidationTask`, and only start using a `TxoValidationTaskV2` when we have two at the same time with something akin to a hard fork where we need to use 2 versions at the same time. If we replace one with the other, I would just replace it keep the name of the replaced version, in this case `TxoValidationTask`

philipr-za commented on base_layer/wallet/src/transaction_service/storage/database.rs:
```diff
@@ -363,6 +431,14 @@ where T: TransactionBackend + 'static
+        self.db.fetch_last_mined_transaction()
+    }
+
+    pub async fn fetch_unmined_transactions(&self) -> Result<Vec<CompletedTransaction>, TransactionStorageError> {
```
perhaps this should be called something like `fetch_unconfirmed_transactions` because it will return mined_unconfirmed which has mined in the name

philipr-za commented on base_layer/wallet/src/transaction_service/protocols/transaction_validation_protocol_v2.rs:
```diff
@@ -0,0 +1,417 @@
+            .await
+            .for_protocol(self.operation_id)?;
+
+        if num_confirmations >= self.config.num_confirmations_required {
```
This feels like a weird Event to use for a lost coinbase that is actually not mined, perhaps a new kind of event for this case?

philipr-za commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -660,10 +672,10 @@ where
-                .set_completed_transaction_validity(tx_id, validity)
-                .await
-                .map(|_| TransactionServiceResponse::CompletedTransactionValidityChanged),
+            /* TransactionServiceRequest::SetCompletedTransactionValidity(tx_id, validity) => self
```
This is used by the output manager to be able to set the validity of Import Transaction's if the output becomes invalid. This is needed because the faux transactions we create to represent imported UTXO's are not checked for validity but should be marked as invalid if the output itself is invalidated.

philipr-za commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -276,9 +275,22 @@ where
             JoinHandle<Result<u64, TransactionServiceProtocolError>>,
         > = FuturesUnordered::new();
 
+        // Should probably be block time or at least configurable
```
Should take into account the power mode timeout that is currently set. These timeouts allow the wallet client to set a much lower interval to reduce polling frequency and save power.

SWvheerden replied:
yeah, a time faster than the block time, does not make sense

philipr-za commented on base_layer/wallet/src/output_manager_service/tasks/txo_validation_task_v2.rs:
```diff
@@ -0,0 +1,420 @@
+    ) -> Result<(), OutputManagerProtocolError> {
+        let unmined_outputs = self
+            .db
+            .fetch_unmined_received_outputs()
```
This doesn't seem correct to me. This function gets all outputs that have `mined_in_block` as null but this includes outputs that are being spent not just received?

Below if these outputs are mined then `update_output_as_mined` is called on them that gives them the Unspent status which seems wrong because they were outgoing funds. 

I see that the next stage of `update_spent_outputs` will correct this but  still the database state at this stage is incorrect.

philipr-za commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -415,7 +459,88 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
+
+    fn mark_output_as_unspent(&self, hash: Vec<u8>) -> Result<(), OutputManagerStorageError> {
+        let conn = self.database_connection.acquire_lock();
+        // Only allow updating of non-deleted utxos
```
not sure what this comment means?

SWvheerden commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db.rs:
```diff
@@ -954,6 +1077,13 @@ struct OutputSql {
+    mined_mmr_position: Option<i64>,
+    marked_deleted_at_height: Option<i64>,
+    marked_deleted_in_block: Option<Vec<u8>>,
+    received_in_tx_id: Option<i64>,
```
should this not be an i64?
You should always receive an output before you you have it, so you should receive it in some transaction. 
Or is this for the faucets and imported? If that's the case cant we add a special recive_id if they don't have a matching id?

SWvheerden commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -1481,30 +1493,39 @@ where
         trace!(target: LOG_TARGET, "Starting transaction validation protocols");
         let id = OsRng.next_u64();
-        let timeout = match self.power_mode {
+        let _timeout = match self.power_mode {
```
remove?

SWvheerden commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -1901,39 +1929,43 @@ where
         }
 
-        let block_height = if let Some(bh) = completed_tx.coinbase_block_height {
+        let _block_height = if let Some(bh) = completed_tx.coinbase_block_height {
```
remove?

SWvheerden commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -1901,39 +1929,43 @@ where
         };
 
-        let timeout = match self.power_mode {
+        let _timeout = match self.power_mode {
```
remove?

SWvheerden commented on comms/src/protocol/rpc/client.rs:
```diff
@@ -226,9 +226,9 @@ impl RpcClientConfig {
     fn default() -> Self {
         Self {
-            deadline: Some(Duration::from_secs(30)),
+            deadline: Some(Duration::from_secs(38)),
```
reason for the 8 sec increase?

SWvheerden commented on comms/src/protocol/rpc/client.rs:
```diff
@@ -226,9 +226,9 @@ impl RpcClientConfig {
+            deadline: Some(Duration::from_secs(38)),
             deadline_grace_period: Duration::from_secs(10),
-            handshake_timeout: Duration::from_secs(30),
+            handshake_timeout: Duration::from_secs(33),
```
and the 3 here?

SWvheerden commented on base_layer/wallet/src/output_manager_service/tasks/txo_validation_task_v2.rs:
```diff
@@ -0,0 +1,420 @@
+            for output in batch {
+                if deleted_bitmap_response
+                    .deleted_positions
+                    .contains(&output.mined_mmr_position.unwrap())
```
can these unwraps fail?

SWvheerden commented on base_layer/wallet/src/output_manager_service/tasks/txo_validation_task_v2.rs:
```diff
@@ -0,0 +1,420 @@
+                if deleted_bitmap_response
+                    .not_deleted_positions
+                    .contains(&output.mined_mmr_position.unwrap()) &&
+                    output.marked_deleted_at_height.is_some()
```
the word deleted here makes this difficult to follow. 
Am I correct in saying this case, the base_node said this utxo is still unspent, it has a mined_mmr height and the wallet thinks this utxo was spent at some height?

SWvheerden commented on base_layer/wallet/src/output_manager_service/tasks/txo_validation_task_v2.rs:
```diff
@@ -0,0 +1,420 @@
+                .for_protocol(self.operation_id)?;
+
+            for output in batch {
+                if deleted_bitmap_response
```
this if checks if the base_node said this utxo was mined at x mmr height but was spent.

SWvheerden commented on base_layer/wallet/src/output_manager_service/tasks/txo_validation_task_v2.rs:
```diff
@@ -0,0 +1,420 @@
+            }
+        }
+
+        loop {
```
wont this be more efficient to rather do this in a single loop. 
We ask for both the mined and unmined from the db.
We then ask for the block which is the furthest back and we repeat this till we found a height that the base_node still has.

philipr-za reviewed (commented):
So I like this approach to doing the validation. By adding some more info about the chain state when outputs and transactions are updated we are able to reduce the complexity of the polling and assumptions around chain state by quite a bit. It also means these calls can be batched quite nicely which will be a major load off.

- I do think we need to be smarter about the interval at which the validation is done. At the very least the interval should be configurable and linked to the power mode so that clients can choose to have a less frequent base node polling

## Output Validation:
- The output validation will set a pending output to be spendable and not encumbered without taking into account the confirmation time
- The get_balance call is still using the `tx_id` field and the PendingTransactionOutputs encumberance mechanism
- cancellation also uses the deprecated `tx_id` field
- Output validation is performed for all outputs including CancelledInbound which is a DoS vector as you could send a wallet many transactions and cancel them and the wallet will now include them in all batch validations.
- Seems like we should remove the pending_transaction_outputs table completely and rather make a new status for pending coinbases
- should remove the concept of Invalid outputs because they are now either spent, unspent or encumbered

## Transaction Validation:
- Should create and emit events for when reorgs occur that can cause a transaction's state to change
- If there are no active transactions we should reduce the frequency of these validations significantly
- PR doesn't invalidate imported Faux transactions when the imported output is invalidated

It is a WIP PR but lots of cleanup needed

SWvheerden commented on base_layer/wallet/src/transaction_service/protocols/transaction_validation_protocol_v2.rs:
```diff
@@ -0,0 +1,417 @@
+        event_publisher: TransactionEventSender,
+    ) -> Self {
+        Self {
+            operation_id: 122, // Get a real tx id
```
What is this used for, tracking?
if so cant we just use a random number here?

SWvheerden commented on base_layer/wallet/src/transaction_service/storage/sqlite_db.rs:
```diff
@@ -1012,7 +1012,7 @@ impl InboundTransactionSql {
         if num_updated == 0 {
             return Err(TransactionStorageError::UnexpectedResult(
-                "Database update error".to_string(),
+                "Updating inbound transactions failed. No rows were affected".to_string(),
```
why not use `num_rows_affected_or_not_found` here

SWvheerden reviewed (commented):
Looks good so far. I tried at the beginning to mark all the commented out and unimplemented so that we don't forget about them, but there is still too many so I stopped that halfway through, Those comments can be disregarded.

hansieodendaal reviewed (commented):
This review focussed on system-level behaviour. 

_General observations are listed below:_
- Mini stress test to three receiving wallets of 2000 transactions each:
  - All transactions completed successfully in the receiving wallets
  - All transactions detected as mined confirmed in the receiving wallets

  **Update 2021/08/19 14:34**
- Mini stress test from three sending wallets of 2000 transactions each:
  - All transactions completed successfully in the sending wallets
  - All transactions detected as mined confirmed in the sending wallets

_Noteworthy findings are listed below:_
- Some transaction monitoring process steps are repeated unnecessarily after successful submission to the unconfirmed mempool, draining resources:
  - `Updated transaction ____ as unmined`
  - `Wallet Event Monitor received wallet transaction service event TransactionBroadcast`
  - `Updating transaction ____ as mined`
  - ` Wallet Event Monitor received wallet transaction service event TransactionMined`
- The wallet consumes a lot of CPU resources, >1x full-time CPU thread, when apparently doing nothing, noted to continue long after many transactions have been received.

  **Update 2021/08/19 14:34**
- I could not pick up any transaction events in the log files, i.e. messages containing `event for tx_id`. It seems the transaction service event stream is broken or publishing events are unreachable in the code (`async fn update_transaction_as_mined(`).
- Log message `trace!(target: LOG_TARGET, "Transaction Validation protocol has ended with result {:?}", join_result);` obfuscates warnings and errors; those should be logged as warning or errors:
  - `TRACE Transaction Validation protocol has ended with result Ok(Ok(122))`
  - `TRACE Transaction Validation protocol has ended with result Ok(Err(TransactionServiceProtocolError { id: 122, error: ConnectivityError { source: ConnectionFailed(ConnectFailedMaximumAttemptsReached) } }))`
  - `TRACE Transaction Validation protocol has ended with result Ok(Err(TransactionServiceProtocolError { id: 122, error: RpcError(HandshakeError(TimedOut)) }))`
  - The console wallet transaction statuses are not always updated, e.g. it shows completed when the log shows it is mined.

Commit:
ci: add pr title check

Commit:
Remove OpenSSL from Windows runtime

- Removed OpenSSL installers and dependencies from Windows runtime
- Removed stdout as appender from console wallet's log4rs logger

Commit:
an alternative transaction validation approach

sdbondi committed:
fix: exit command and free up tokio thread (#3235)

Description
---
Exit command didn't exit the cli loop.
Rustyline was holding up a tokio thread - used a blocking thread instead

Motivation and Context
---
Bug

How Has This Been Tested?
---
Manually on base node

aviator-app[bot] committed:
Merge branch 'development' into gha-pr-title

Commit:
fix: add status output to logs in non-interactive mode (#3244)

Description
---
Add status output to logs in non-interactive mode

Motivation and Context
---
Base node in non-interactive mode was logging status

How Has This Been Tested?
---
Manually run base node in non-interactive mode

aviator-app[bot] committed:
Merge branch 'development' into gha-pr-title

Cifko committed:
feat: add tab for error log to the wallet

Commit:
ci: add pr title check (#3233)

Description
---

Add a github action to check PR titles follow conventional commit spec

Motivation and Context
---

Automate checks

How Has This Been Tested?
---

github actions

Commit:
fix: remove OpenSSL from Windows runtime (#3242)

Description
---
- Removed OpenSSL installers and dependencies from Windows runtime
- Removed stdout as appender from console wallet's log4rs logger

Motivation and Context
---
- OpenSSL is not a Windows runtime dependency anymore
- The console wallet cannot handle log4rs logger messages to its stdout as it breaks the TUI

How Has This Been Tested?
---
Build the installer, ran an installer, ran all the executable components

Commit:
Merge branch 'development' into wallet-error-window

Commit:
feat(wallet): add tab for error log to the wallet (#3250)

Description
---
Add a new tab in the console wallet. In the new tab is a stdout log from the wallet. The stdout is redirected to a new logfile called "stdout.log"
I've added also some coloring, it's much easier to read the log file with some simple colors.

Motivation and Context
---
Easier to debug the wallet.

How Has This Been Tested?
---
Manually.

Commit:
reformat SQL code

Cifko committed:
feat: add support for forcing sync from seeds (#3228)

## Description
Add support for forcing sync to a seed node. Specify index for the node from peer_seeds list.

## How Has This Been Tested?
Manually.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I have squashed my commits into a single commit.

Cifko committed:
test: cucumber forced sync (#3230)

## Description
Add cucumber test that tests forced sync to single node.

## How Has This Been Tested?
npm test -- --name "Force sync many nodes agains one peer"

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I have squashed my commits into a single commit.

StriderDM committed:
test: add multiple wallet recovery from peer (#3240)

Description
Added multiple wallet recovery from peer

Motivation and Context
Additional tests

How Has This Been Tested?
Manually (npm test -- --name "Multiple Wallet recovery from seed node")

StriderDM committed:
feat: base_node prompt user to create id if not found (#3245)

Description
Prompt user to create id if not found

Motivation and Context
Improvement to base node startup, specifically on first run.

How Has This Been Tested?
Manually

sdbondi committed:
fix: daily wallet recovery fixes (#3229)


<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
- Use the GRPC client connection given `WalletProcess`
- Outputs error details in webhook in recovery cron job
- Adds very basic mocha tests

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Recovery daily is failing even though it succeeds. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I have squashed my commits into a single commit.

philipr-za committed:
feat!: tell an FFI client that a recovery is in progress on `wallet_create` (#3249)

Description
---
### Note: This is a breaking change to LibWallet FFI

Currently if a wallet recovery was in progress and the wallet was shutdown the next time that wallet is start by an FFI client using the ‘wallet_create’ method there is no way for the FFI client to know that the recovery should be continued.

The wallet is able to resume the recovery from where it left off and it should so as not to lose funds but the FFI client must restart the recovery process with the same seed words. The FFI client has to do the restarting so that it can provide the callback through which the process is monitored.

Furthermore, the wallet does not respond to P2P transaction negotiation message if a recovery process is in progress so it is important that an FFI client completes any outstanding recoveries ASAP.

How Has This Been Tested?
---
untested in the backend.

hansieodendaal committed:
fix: fix base_node_service_config not read (#3251)

Description
---
Fixed the base_node_service_config not being initialized with values from the config file.

Motivation and Context
---
See above

How Has This Been Tested?
---
System level testing

Commit:
Merge branch 'development' into mb-transaction-validation

Commit:
wip
